Add a ci-operator Prow job generator

petr-muller opened the pull request:
Add an utility to generate Prow job configuration from ci-operator configuration files.

/assign @stevekuznetsov @bbguimaraes @droslean 

The purpose of this tool is to reduce an amount of boilerplate that component owners need to write when they use ci-operator to set up CI for their component. The generator is able to entirely generate the necessary Prow job configuration from the ci-operator configuration file.

**Full-repository**
The generator can use the naming convention and directory structure of this repository. Using this mode, all you need to do is to place your ci-operator configuration file to the correct place in ci-operator/config directory and then run the generator with `--full-repo` parameter. With `--full-repo` option, the generator uses git to detect a root directory of the repo and then it constructs paths to both directories from it. It is also possible to not use the auto-detection and pass the paths directly.

**Single configuration**
You can use `--source-config` option instead to pass a single ci-operator configuration file. In this case, the generator will print the Prow job config YAML to the standard output.

**What does the generator create**

The generator creates one presubmit and one postsubmit job for each test specified in the ci-operator config file (in `tests` list). Also, if the configuration file has a non-empty `images` list, one additional presubmit and postsubmit job is generated with `--target=[images]` option passed to ci-operator to attempt to build the component images. This postsubmit job also uses the `--promote` option to promote the component images built in this way.

**TODO**
- [x] Postsubmit jobs
- [x] Prune `null`s and empty strings?
- [x] Sort out the type mess (test-infra aliases vs kube/api types)
- [x] Non-test targets (`[images]`?)
- [x] Inject jobs into an existing YAML config instead of stdout
- [x] Recursively walk over all ci-operator configs and generate all Prow configs?
- [x] `labels` field for the `-images` postsubmit job
- [x] Tests
- [x] Documentation
- [x] Fix paths in README
- [x] Integration source->target test
- [x] Remove the [images] hack
- [x] Include branch in the job name so branch jobs do not have identical names.
- [x] Remove comments
- [x] Address the nits in last Steve's review
- [x] Depython the tests
- [x] Handle file path element extraction error cases better

petr-muller commented:
One thing that puzzles me is that I am constructing a `PodSpec` in [generatePodSpec](https://github.com/openshift/release/pull/1160/files#diff-6830cc770f52feabaaf7e44046a8bf73R34). The instance should ultimately be assigned into the `Presubmit` instance member provided by `test-infra/prow/config`, which does not use types from `api/core/v1` ([here](https://github.com/kubernetes/api/blob/master/core/v1/types.go)) but their aliases defined in `test-infra` [here](https://github.com/kubernetes/test-infra/blob/master/prow/kube/types.go#L55).

One piece of the type hierarchy that is missing in `k8s.io/test-infra/prow/kube` is `EnvVarSource`. What is the better route? Add `EnvVarSource` alias there, or use the types straight from `k8s.io/api` and somehow cast/type convert (is that even possible?) so that I can assign it to the `Presubmit` instance?

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
```
nit: we group imports with stdlib/external/internal like

```go
import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"strings"
	
	"github.com/ghodss/yaml"
	cioperatorapi "github.com/openshift/ci-operator/pkg/api"
	prowconfig "k8s.io/test-infra/prow/config"
	prowkube "k8s.io/test-infra/prow/kube"
)
```

petr-muller replied:
Aha, blank line makes the formatter do what I want!

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,126 @@
+	return opt
+}
+
+func generatePodSpec() *prowkube.PodSpec {
```
We are going to migrate to the "real" upstream `PodSpec` (from k8s api, not `test-infra`) soon. There is no reason other than historical silliness as to why the `test-infra` type is different. You should be OK to target the upstream type here and `test-infra` will catch up soon.

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,126 @@
+
+func generatePresubmitForTest(test, org, component, branch string) prowconfig.Presubmit {
+	presubmit := prowconfig.Presubmit{Agent: "kubernetes", AlwaysRun: true}
+	presubmit.Name = fmt.Sprintf("ci-operator-%s-%s-%s", org, component, test)
```
You can add these to the literal declaration instead of doing assignment for clarity

petr-muller replied:
Hehe, I struggled how to write a multi-line initializer, so I did it this way. I found out how to do this now.

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,126 @@
+	return &podspec
+}
+
+func generatePresubmitForTest(test, org, component, branch string) prowconfig.Presubmit {
```
We commonly use `repo` instead of `component`

petr-muller replied:
I used `repo` for `organization/repo` string already. Does that have common name? I used `orgrepo` for now and changed `component` to `repo`.

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,126 @@
+	presubmit.Brancher = prowconfig.Brancher{}
+	presubmit.Brancher.Branches = []string{branch}
+	presubmit.UtilityConfig = prowconfig.UtilityConfig{Decorate: true}
+	// TODO: `skip_cloning` does not seem to be covered by `Presubmit`
```
Should be under `DecorationConfig.SkipCloning`

petr-muller replied:
Indeed!

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,126 @@
+
+func generatePresubmits(tests []cioperatorapi.TestStepConfiguration, org, component, branch string) map[string][]prowconfig.Presubmit {
+	repo := fmt.Sprintf("%s/%s", org, component)
+	presubmits := make(map[string][]prowconfig.Presubmit)
```
nit: don't need to explicitly `make` in Go (often), this could be:

```go
presubmits := map[string][]prowconfig.Presubmit{}
for _, test := range tests {
    presubmits[repo] = append(presubmits[repo], ...)
}
```

stevekuznetsov commented:
@petr-muller wait, isn't the definition [here](https://github.com/kubernetes/test-infra/blob/master/prow/config/jobs.go#L118) using the correct upstream type? Where is the internal alias?

petr-muller commented:
@stevekuznetsov uh, I wonder how I got mislead :/

stevekuznetsov commented:
Maybe you were looking at a different commit? What you are saying was definitely true in the recent past

petr-muller commented:
Addressed review comments. I am not able to build the generator after using the `k8s.io/api/core/v1` types because of type mismatch which I am not able to solve on my own:

```
./main.go:71:7: cannot use generatePodSpec(org, repo, branch) (type *"k8s.io/api/core/v1".PodSpec) as type *"k8s.io/test-infra/vendor/k8s.io/api/core/v1".PodSpec in field value
```

petr-muller commented:
For now, I am working around the dependency / type issues by using the type aliases in `test-infra`. I am maintaining a branch on top of this PR with that modification: https://github.com/petr-muller/release/commits/prowgen-prowkube. To make it work, I also needed to add few aliases to `test-infra`, which I track (https://github.com/petr-muller/test-infra/commit/b4b2ae98c89baffa14f2cd774f402ae1b8e4695f)

petr-muller commented:
The issue I am seeing right now: I can build the generator now, but after I execute it I get:

```
$  ./ci-operator-prowgen 
./ci-operator-prowgen flag redefined: log_dir
panic: ./ci-operator-prowgen flag redefined: log_dir

goroutine 1 [running]:
flag.(*FlagSet).Var(0xc4200aa060, 0x101c460, 0xc4200b7910, 0xf637d9, 0x7, 0xf84a31, 0x2f)
	/usr/lib/golang/src/flag/flag.go:810 +0x540
flag.(*FlagSet).StringVar(0xc4200aa060, 0xc4200b7910, 0xf637d9, 0x7, 0x0, 0x0, 0xf84a31, 0x2f)
	/usr/lib/golang/src/flag/flag.go:713 +0x8b
flag.(*FlagSet).String(0xc4200aa060, 0xf637d9, 0x7, 0x0, 0x0, 0xf84a31, 0x2f, 0xc4200b7900)
	/usr/lib/golang/src/flag/flag.go:726 +0x8b
flag.String(0xf637d9, 0x7, 0x0, 0x0, 0xf84a31, 0x2f, 0xc42018d800)
	/usr/lib/golang/src/flag/flag.go:733 +0x69
```

stevekuznetsov commented:
I use `dep` for vendoring -- look at `ci-vm-operator/Gopkg.toml` for the declarative file, once that is written you can simply use `dep ensure` in your repo to solve versions and vendor code.

stevekuznetsov commented:
More insight after mucking with config today -- I think we can special-case a `--target=[images]` presubmit and a `--target=[images] --promote` postsubmit for all repos?

/cc @smarterclayton

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,236 @@
+	imagesTest := false
+
+	for _, element := range configSpec.Tests {
+		// Check if config file has "images" test defined to avoid name clash
```
NVM figured it out

petr-muller replied:
I'm taking of the special case where someone would have `images` test in their config file. The generator adds a job with `--target=[images]` parameter in addition to the jobs derived from `tests` keys. The generator derives names from the test identifier, but the job name cannot be `...-[images]` so the generator calls such job `...--images` which would clash with a job made for `images` test in the config file.

But... while writing this, I realized - is it really the case? Maybe I was just too careful and it is not actually needed. @stevekuznetsov, do you think we could simply use `[images]` there? It would mean we have job called `...-[images]` and stuff like `context: ci/prow/[images]` and `rerun_command: /test [images]`. It would make the generator code simpler.

stevekuznetsov replied:
I think even if it is allowed, hiding the `[` on `images` should be something we do, it's cleaner for the end user. The `[images]` and promotion tasks are more admin tasks than tests so we can own the generation of these 100% in this tool.

smarterclayton commented:
I think that would be right.  Also, the target images post submit should have the correct prow job labels (the ones in config) for artifacts: true / etc so we can filter on that when we want to "rebuild all images associated with the master branch".

petr-muller commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,236 @@
+			Decorate:         true,
+		},
+	}
+	presubmit.Spec.Containers[0].Args = append(
```
We can. I think I had some reason but it apparently got lost during refactorings.

petr-muller commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,236 @@
+}
+
+// these are unnecessary, and make the config larger so we strip them out
+func yamlBytesStripNulls(yamlBytes []byte) []byte {
```
Sounds good. I'll make a PR there and remove this hack if accepted.

petr-muller replied:
https://github.com/kubernetes/test-infra/pull/9024

petr-muller commented:
> More insight after mucking with config today -- I think we can special-case a --target=[images] presubmit and a --target=[images] --promote postsubmit for all repos?

I am already adding the `--target=[images]` job to both presubmit and postsubmit, but currently *all* postsubmits have `--promote` which probably is not correct. I will fix it so that only the `[images]` job has the `--promote`.

petr-muller commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,236 @@
+}
+
+func generateAllProwJobs() {
+	repoRootRaw, err := exec.Command("git", "rev-parse", "--show-toplevel").Output()
```
I was thinking that this is really a convenience tool and using it should be as simple as possible. The generator will almost always be called inside this repo and because we have a standard directory structure, everyone will always pass the same directories there. We can compute these defaults easily, why not do so?

I can of course add options to pass these manually, but I think we should keep the "easy" way, too.

petr-muller commented:
I've added vendorized dependencies and now I can build and execute the thing.

stevekuznetsov commented:
Cool! Can we generate and look at diffs? Let's also write some unit tests!

petr-muller commented:
I have rebased everything to get rid of the sausage, added unit-tests and a README. I think the generator is now feature complete for a first submission. I will now generate some diffs against the current config files to see how the generated jobs differ from whatever humans configured.

stevekuznetsov commented on tools/ci-operator-prowgen/README.md:
````diff
@@ -0,0 +1,140 @@
+To use the generator, you need to build it:
+
+```
+$ go build
````
From root of repo?

```
go build tools/ci-operator-prowgen/...
```

petr-muller replied:
I thought it would be better for the snippets to match the README location but OK.

stevekuznetsov commented on tools/ci-operator-prowgen/README.md:
````diff
@@ -0,0 +1,140 @@
+with `--full-repo` parameter:
+
+```
+$ ./ci-operator-prowgen --full-repo
````
When we shove this into a container to run it, we add a dependency on `git`. Not sure it's super useful in addition to the explcit dir-as-flags approach if we can just add a script to this repo to run the generator with the appropriate dirs, but I guess if we ensure we thread through the dependencies as necessary it should be fine either way

petr-muller replied:
I think this is a question of main use case - right now, do we intend the generator to be used by humans or do we want it to be shoved into some automated script? If the former then I would really like to keep it (not fond of wrapping stuff in scripts so it's usable).

stevekuznetsov commented on tools/ci-operator-prowgen/README.md:
```diff
@@ -0,0 +1,140 @@
+expected to have a `(anything)/ORGANIZATION/REPO/BRANCH.extension` form, just
+likes path in [ci-operator/config](../..ci-operator/config) do.
+
+## What does the generator create
```
It's convention upstream to create a file per type of job per owner (cross product of pre/postsubmit, periodic and org, repo, branch). Would it be hard to do that with this tool, too?

petr-muller replied:
Hmm. Generating pre/post submits separately is easy. Generating pre/submits from multiple sources into a single file would be slightly more work, but not super-hard. I think the best way how to do this would be to implement the ability to merge into an existing file, but I also foresee there's a little can of worms of complexity hiding there, which I wanted to avoid for a initial submission. 

To be clear,  for the input directory of this:

```
.../ci-operator/config/openshift/component/master.json
.../ci-operator/config/openshift/component/release-branch.json
```

do you want to generate this (trivial):
```
.../ci-operator/jobs/openshift/component/presubmits-master.yaml
.../ci-operator/jobs/openshift/component/presubmits-release-branch.yaml
.../ci-operator/jobs/openshift/component/postsubmits-master.yaml
.../ci-operator/jobs/openshift/component/postsubmits-release-branch.yaml
```

or this (not trivial):
```
.../ci-operator/jobs/openshift/component/presubmits.yaml # has jobs for both master and release-branch
.../ci-operator/jobs/openshift/component/postsubmits.yaml # ditto
```
?

stevekuznetsov replied:
Adding branch to the name is not really necessary, we shadow names upstream in that case. But if we do it, it's fine too.

stevekuznetsov commented on tools/ci-operator-prowgen/README.md:
````diff
@@ -0,0 +1,140 @@
+The generator creates one presubmit and one postsubmit job for each test
+specified in the ci-operator config file (in `tests` list):
+
+```yaml
````
Would be cool if we had an integration test with a sample JSON and a sample YAML and a test that verified that one was generated from the other

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,318 @@
+// Generate a PodSpec that runs `ci-operator`, to be used in Presubmit/Postsubmit
+// Various pieces are derived from `org`, `repo`, `branch` and `target`.
+// `additionalArgs` are passed as additional arguments to `ci-operator`
+// Example
```
I'd say let's not add these as the comments will rot fairly quick and we can achieve examples that can't rot through tests

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,236 @@
+	Target string
+}
+
+func generatePresubmitForTest(test testDescription, org, repo, branch string) *prowconfig.Presubmit {
```
Yep, that was my read too

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,318 @@
+			Decorate:         true,
+		},
+	}
+	return &postsubmit
```
you can also just 

```go
return &prowconfig.Postsubmit{
```

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,318 @@
+		// (we generate the additional `--target=[images]` jobs name with `images`
+		// as an identifier, but a user can have `images` test defined in his
+		// config file which would result in a clash)
+		if element.As == "images" {
```
Which config files have an explicit images target like this?

stevekuznetsov replied:
(maybe we can just remove that)

petr-muller replied:
None. But that word is not that uncommon and once someone calls his test like this, the generator would *silently* not generate something that it would otherwise, which is a frustrating error.

stevekuznetsov replied:
I think we should add validation to the `ci-operator` that disallows such a test target name

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,318 @@
+	data, err := ioutil.ReadFile(configFilePath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to read ci-operator config (%v)\n", err)
+		os.Exit(1)
```
idiomatic nit: Commonly errors are bubbled up to the top-level `main()` which is the only part to use `os.Exit()`. More useful if we're breaking this into multiple files, but might be worth doing now anyway

petr-muller replied:
OK

stevekuznetsov commented on tools/ci-operator-prowgen/main_test.go:
```diff
@@ -0,0 +1,173 @@
+}
+
+func TestGeneratePodSpec(t *testing.T) {
+	podSpec := generatePodSpec("organization", "repo", "branch", "target")
```
Generally you should be able to use `reflect.DeepEqual` for comparison of structs. However, there are semantic meanings to fields in `PodSpec` that exist on top of the normal equality rules for e.g. lists or maps, so we need to use the k8s equality library. We can also use the k8s diff library to get nice diffs on these when they are wrong. See this example: https://github.com/kubernetes/test-infra/blob/master/prow/pod-utils/decorate/podspec_test.go#L711-L713

petr-muller replied:
OK

stevekuznetsov commented on tools/ci-operator-prowgen/main_test.go:
```diff
@@ -0,0 +1,173 @@
+	)
+}
+
+func TestGeneratePodSpecWithAdditionalArgs(t *testing.T) {
```
Idiom in Go is to use a `for` loop and sub-tests for similar tests like this https://blog.golang.org/subtests

petr-muller replied:
OK

stevekuznetsov commented on tools/ci-operator-prowgen/main_test.go:
```diff
@@ -0,0 +1,173 @@
+	checkCondition(t, "len(presubmits) == 1", len(presubmits) == 1)
+	checkCondition(t, "len(postsubmits) == 1", len(postsubmits) == 1)
+	checkCondition(t,
+		"len(presubmits[\"organization/repo\"]) == 2",
```
Will be not important if you use the `quality.Semantic.DeepEqual` as mentioned above, but FWIW in Go you can use back-tick quotes to have quote literals without escaping:

    `len(presubmits["organization/repo"]) == 2`,

petr-muller replied:
OK, thanks

stevekuznetsov commented on tools/ci-operator-prowgen/main.go:
```diff
@@ -0,0 +1,318 @@
+// about component repository information.
+// The convention for ci-operator config files in this repo:
+// ci-operator/config/ORGANIZATION/COMPONENT/BRANCH.json
+func extractRepoElementsFromPath(configFilePath string) (string, string, string) {
```
What happens in error cases?

petr-muller replied:
This should happen only when someone passes some funky path to `--source-config`, which we warn against in [README](https://github.com/openshift/release/pull/1160/commits/25065fd8a9bea36569b15f1517bf07c3cc93268a#diff-35a06896ff3190034afcf280abb5efbaR57) - we only write to stdout so at least we do not write some crap to job directories. Or someone passes a path to `--config-dir ../../ci-operator/config/` where that path has some different directory structure but still contains ci-operator config files. I think that someone sort of knows what they are doing and can verify what was generated.

Good case, those fragments will not be there, we can detect (we do not now) it and error out.
Bad case, those fragments will be there but will be something bogus, and we will generate some crap. We could only detect this by asking github whether that org/repo:branch triplet is sane but IMO that is overkill.

petr-muller commented:
The generator now generates separate -presubmit and -postsubmit files, following the structure that is already in this repo. It merges the generated content with existing ones (jobs are matched by name).

I have tried to apply the generator and commited the result here: https://github.com/petr-muller/release/commit/2453ff2bffce179bb31d56e54bd9475810010976

Unfortunately the diff does not give very useful output because the ordering of YAML elements is so different.

stevekuznetsov commented:
Used the generator today, this looks really good. On scrum you mentioned the last blocker was an integration test for this tool, then it is ready? Some follow-up items for when this merges:

 - we need to add validation to `ci-operator` so that nobody can name a test target "images"
 - we need to document how to use the tool and update the onboarding doc
 - we need to do your read/no-op/write commit to reformat everything
 - repo by repo, we should (re)generate the Prow config and use a tool to generate a human-readable digested diff, maybe trying to explain what has changed for the user -- are any `/test foo` incantations different, etc
 - repo by repo, these changes should be created into a PR so we can farm out reviews

Excited to see this merge!

smarterclayton commented:
Steve asked me to comment here:

1. Don't run the same postsubmits that are presubmits.  The merge queue is already required to run them before we merge, and no one looks at post submits, so we're going to catch any regressions or breaks when the PR job breaks next.
2. Only images that are part of a release should have release_tag_configuration or have the prow label artifact: images.  These are used to assemble sets of coherent images - things like ci-operator or other tool images don't become part of our release config.

smarterclayton commented on tools/ci-operator-prowgen/Gopkg.toml:
```diff
@@ -3,5 +3,7 @@ required = [
 	"k8s.io/test-infra/prow/config",
 	"k8s.io/test-infra/prow/kube",
-	"k8s.io/api/core/v1"
+	"k8s.io/api/core/v1",
```
I'm confused why these are necessary.  I would not expect 5 million lines of vendor dependencies to be pulled in just for one test case.

smarterclayton commented:
I'd prefer not to have to vendor the entire kube directory.  5 million lines of code added to openshift/release is too much.  If we really need those dependencies we should just move this to openshift/prow-config-generator or whatever.

petr-muller commented:
All comments were addressed. The generator now generates only presubmits for tests, plus an additional presubmit + postsubmit for `--target=[images]`. The images postsubmit no longer has `artifacts: images` label: we'll need to figure out how to detect whether a job should have it or not in a future fix. I have removed the vendored code and committed just the `Gopkg.*` files for now.

stevekuznetsov commented:
/lgtm

stevekuznetsov commented:
More follow-ups:
 - fix vendoring issues
 - detect if promotion is to origin imagestreams to add the label
 - all above listed ones
 - add a presubmit and whitelist of repos to use on it

stevekuznetsov commented:
@petr-muller this might also be a pre-requisite
